external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala

```diff
@@ -144,4 +150,174 @@ object KafkaUtils {
+   */
+  @Experimental
+  def createRDD[
+    K: ClassTag,
+    V: ClassTag,
+    U <: Decoder[_]: ClassTag,
+    T <: Decoder[_]: ClassTag,
+    R: ClassTag] (
+      sc: SparkContext,
+      kafkaParams: Map[String, String],
+      offsetRanges: Array[OffsetRange],
+      leaders: Array[Leader],
+      messageHandler: MessageAndMetadata[K, V] => R
+  ): RDD[R] with HasOffsetRanges = {
+
+    val leaderMap = leaders
+      .map(l => TopicAndPartition(l.topic, l.partition) -> (l.host, l.port))
+      .toMap
+    new KafkaRDD[K, V, U, T, R](sc, kafkaParams, offsetRanges, leaderMap, messageHandler)
+  }
+
+  /**
+   * This stream can guarantee that each message from Kafka is included in transformations
+   * (as opposed to output actions) exactly once, even in most failure situations.
+   *
+   * Points to note:
+   *
+   * Failure Recovery - You must checkpoint this stream, or save offsets yourself and provide them
+   * as the fromOffsets parameter on restart.
+   * Kafka must have sufficient log retention to obtain messages after failure.
+   *
+   * Getting offsets from the stream - see programming guide
+   *
+.  * Zookeeper - This does not use Zookeeper to store offsets.  For interop with Kafka monitors
+   * that depend on Zookeeper, you must store offsets in ZK yourself.
+   *
+   * End-to-end semantics - This does not guarantee that any output operation will push each record
+   * exactly once. To ensure end-to-end exactly-once semantics (that is, receiving exactly once and
+   * outputting exactly once), you have to either ensure that the output operation is
+   * idempotent, or transactionally store offsets with the output. See the programming guide for
+   * more details.
+   *
+   * @param ssc StreamingContext object
+   * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
+   * configuration parameters</a>.
+   *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
+   *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+   * @param messageHandler function for translating each message into the desired type
+   * @param fromOffsets per-topic/partition Kafka offsets defining the (inclusive)
+   *  starting point of the stream
+   */
+  @Experimental
+  def createNewStream[
+    K: ClassTag,
+    V: ClassTag,
+    U <: Decoder[_]: ClassTag,
+    T <: Decoder[_]: ClassTag,
+    R: ClassTag] (
+      ssc: StreamingContext,
+      kafkaParams: Map[String, String],
+      fromOffsets: Map[TopicAndPartition, Long],
+      messageHandler: MessageAndMetadata[K, V] => R
+  ): InputDStream[R] = {
+    new DeterministicKafkaInputDStream[K, V, U, T, R](
+      ssc, kafkaParams, fromOffsets, messageHandler)
+  }
+
+  /**
+   * This stream can guarantee that each message from Kafka is included in transformations
+   * (as opposed to output actions) exactly once, even in most failure situations.
```

Comment:
I really think the best way to handle the exactly-once semantics question
isn't in javadoc, but in separate programming guide that covers

For at-most-once: do this...

For at-least-once: do this...

For exactly-once when you have idempotent storage of outputs: do this...

For exactly-once when you have transactional storage of outputs: do this...

For exactly-once when you have neither idempotent nor transactional
storage:  Can't do it, outputs may be duplicated.


Having the javadoc just say some variation on "this does exactly once" is
really oversimplifying the situation IMHO.

On Tue, Feb 3, 2015 at 1:03 PM, Patrick Wendell <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala
> <https://github.com/apache/spark/pull/3798#discussion_r24029286>:
>
> > +    V: ClassTag,
> > +    U <: Decoder[_]: ClassTag,
> > +    T <: Decoder[_]: ClassTag,
> > +    R: ClassTag] (
> > +      ssc: StreamingContext,
> > +      kafkaParams: Map[String, String],
> > +      fromOffsets: Map[TopicAndPartition, Long],
> > +      messageHandler: MessageAndMetadata[K, V] => R
> > +  ): InputDStream[R] = {
> > +    new DeterministicKafkaInputDStream[K, V, U, T, R](
> > +      ssc, kafkaParams, fromOffsets, messageHandler)
> > +  }
> > +
> > +  /**
> > +   * This stream can guarantee that each message from Kafka is included in transformations
> > +   * (as opposed to output actions) exactly once, even in most failure situations.
>
> So what about saying it provides exactly once semantics subject to the
> existing fault recovery mechanisms in streaming (i.e. checkpointing). It
> just conveys basically no information now, it says it might work and might
> not.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r24029286>.
>